alexa/request/certificate.php

```diff
@@ -0,0 +1,185 @@
+
+		// 3. Checking if the certificate is not expired and has the right SAN
+		$this->validate_certificate();
+
+		// 4. Verifying the request signature
+		$this->validaterequest_signature( $request_data );
+
+		// Set http status code back to 200
+		status_header( 200 );
+	}
+
+	/**
+	 * Check if request is whithin the allowed time.
+	 * @throws InvalidArgumentException
+	 */
+	public function validate_timestamp( $timestamp ) {
+		$now = new DateTime;
+		$timestamp = new DateTime( $timestamp );
+		$difference_in_seconds = $now->getTimestamp() - $timestamp->getTimestamp();
+
+		if ( $difference_in_seconds > self::TIMESTAMP_VALID_TOLERANCE_SECONDS ) {
+			throw new InvalidArgumentException( 'Request timestamp was too old. Possible replay attack.' );
+		}
+	}
+
+	public function validate_certificate() {
+		$this->certificate_content = $this->get_certificate();
+		$parsed_certificate = $this->parse_certificate( $this->certificate_content );
+
+		if ( ! $parsed_certificate || ! $this->validate_certificate_date( $parsed_certificate ) || ! $this->validate_certificate_san( $parsed_certificate, static::ECHO_SERVICE_DOMAIN ) ) {
+			throw new InvalidArgumentException( "The remote certificate doesn't contain a valid SANs in the signature or is expired." );
+		}
+	}
+	/*
+	 * @params $request_data
+	 * @throws InvalidArgumentException
+	 */
+	public function validaterequest_signature( $request_data ) {
+		$cert_key = openssl_pkey_get_public( $this->certificate_content );
+		$signature = base64_decode( $this->request_signature );
+
+		$valid = openssl_verify( $request_data, $signature, $cert_key, self::ENCRYPT_METHOD );
+		if ( ! $valid ) {
+			throw new InvalidArgumentException( 'Request signature could not be verified' );
+		}
+	}
+
+	/**
+	 * Returns true if the ceertificate is not expired.
+	 *
+	 * @param array $parsed_certificate
+	 * @return boolean
+	 */
+	public function validate_certificate_date( array $parsed_certificate ) {
+		$valid_from = $parsed_certificate['validFrom_time_t'];
+		$valid_to = $parsed_certificate['validTo_time_t'];
+		$time = time();
+		return ( $valid_from <= $time && $time <= $valid_to );
+	}
+
+	/**
+	 * Returns true if the configured service domain is present/valid, false if invalid/not present
+	 * @param array $parsed_certificate
+	 * @return boolean
+	 */
+	public function validate_certificate_san( array $parsed_certificate, $amazon_service_domain ) {
+		if ( false === strpos( $parsed_certificate['extensions']['subjectAltName'], $amazon_service_domain ) ) {
+			return false;
+		} else {
+			return true;
+		}
+	}
+
+	/**
+	 * Verify URL of the certificate
+	 * @throws InvalidArgumentException
+	 * @author Emanuele Corradini <emanuele@evensi.com>
+	 */
+	public function verify_signature_certificate_url() {
+		$url = parse_url( $this->certificate_url );
+
+		if ( static::SIGNATURE_VALID_PROTOCOL !== $url['scheme'] ) {
+			throw new InvalidArgumentException( 'Protocol isn\'t secure. Request isn\'t from Alexa.' );
```

Comment:
Use double quotes to avoid backslashing here and the lines below. Also same translation feedback.